d2ir/d2ir.go

```diff
@@ -0,0 +1,1046 @@
+	}
+
+	if hasLayerKeywords(head) != -1 && NodeLayerKind(m) == "" {
+		return nil, d2parser.Errorf(kp.Path[i].Unbox(), "%s is only allowed at a layer root", head)
+	}
+
+	for _, f := range m.Fields {
+		if !strings.EqualFold(f.Name, head) {
+			continue
+		}
+
+		// Don't add references for fake common KeyPath from trimCommon in CreateEdge.
+		if refctx != nil {
+			f.References = append(f.References, &FieldReference{
+				String:  kp.Path[i].Unbox(),
+				KeyPath: kp,
+				Context: refctx,
+			})
+		}
+
+		if i+1 == len(kp.Path) {
+			return f, nil
+		}
+		if _, ok := f.Composite.(*Array); ok {
+			return nil, d2parser.Errorf(kp.Path[i].Unbox(), "cannot index into array")
+		}
+		if f.Map() == nil {
+			f.Composite = &Map{
+				parent: f,
+			}
+		}
+		return f.Map().ensureField(i+1, kp, refctx)
+	}
+
+	f := &Field{
+		parent: m,
+		Name:   head,
+	}
+	// Don't add references for fake common KeyPath from trimCommon in CreateEdge.
+	if refctx != nil {
+		f.References = append(f.References, &FieldReference{
+			String:  kp.Path[i].Unbox(),
+			KeyPath: kp,
+			Context: refctx,
+		})
+	}
+	m.Fields = append(m.Fields, f)
+	if i+1 == len(kp.Path) {
+		return f, nil
+	}
+	f.Composite = &Map{
+		parent: f,
+	}
+	return f.Map().ensureField(i+1, kp, refctx)
+}
+
+func (m *Map) DeleteField(ida ...string) *Field {
+	if len(ida) == 0 {
+		return nil
+	}
+
+	s := ida[0]
+	rest := ida[1:]
+
+	for i, f := range m.Fields {
+		if !strings.EqualFold(f.Name, s) {
+			continue
+		}
+		if len(rest) == 0 {
+			m.Fields = append(m.Fields[:i], m.Fields[i+1:]...)
+			return f
+		}
+		if f.Map() != nil {
+			return f.Map().DeleteField(rest...)
+		}
+	}
+	return nil
+}
+
+func (m *Map) GetEdges(eid *EdgeID) []*Edge {
+	eid, m, err := eid.resolveUnderscores(m)
+	if err != nil {
+		return nil
+	}
+	common, eid := eid.trimCommon()
+	if len(common) > 0 {
+		f := m.GetField(common...)
+		if f == nil {
+			return nil
+		}
+		if f.Map() != nil {
+			return f.Map().GetEdges(eid)
+		}
+		return nil
+	}
+
+	var ea []*Edge
+	for _, e := range m.Edges {
+		if e.ID.Match(eid) {
+			ea = append(ea, e)
+		}
+	}
+	return ea
+}
+
+func (m *Map) CreateEdge(eid *EdgeID, refctx *RefContext) (*Edge, error) {
+	if ParentEdge(m) != nil {
+		return nil, d2parser.Errorf(refctx.Edge, "cannot create edge inside edge")
+	}
+
+	eid, m, err := eid.resolveUnderscores(m)
+	if err != nil {
+		return nil, d2parser.Errorf(refctx.Edge, err.Error())
+	}
+	common, eid := eid.trimCommon()
+	if len(common) > 0 {
+		tmp := *refctx.Edge.Src
+		kp := &tmp
+		kp.Path = kp.Path[:len(common)]
+		f, err := m.EnsureField(kp, nil)
+		if err != nil {
+			return nil, err
+		}
+		if _, ok := f.Composite.(*Array); ok {
+			return nil, d2parser.Errorf(refctx.Edge.Src, "cannot index into array")
+		}
+		if f.Map() == nil {
+			f.Composite = &Map{
+				parent: f,
+			}
+		}
+		return f.Map().CreateEdge(eid, refctx)
+	}
+
+	ij := hasLayerKeywords(eid.SrcPath...)
```

Comment:
i guess we should just disallow any reserved keyword (i.e. each part of edge endpoint has a shape field)

Reply:
Oh true.